src/sage/rings/polynomial/polynomial_ring.py

```diff
@@ -1033,6 +1034,75 @@ def __hash__(self):
         h = self._cached_hash = hash((self.base_ring(),self.variable_name()))
         return h
 
+    def __iter__(self):
+        r"""
+        Return iterator over the elements of this polynomial ring.
+
+        EXAMPLES::
+
+            sage: from itertools import islice
+            sage: R.<x> = GF(3)[]
+            sage: list(islice(iter(R), 10))
+            [0, 1, 2, x, x + 1, x + 2, 2*x, 2*x + 1, 2*x + 2, x^2]
+
+        TESTS::
+
+            sage: R.<x> = Integers(1)[]
+            sage: [*R]
+            [0]
+            sage: R.<x> = QQ[]
+            sage: list(islice(iter(R), 10))
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: iteration over infinite base ring not yet implemented
+        """
+        # adapted from sage.modules.free_module.FreeModule_generic.__iter__
+        R = self.base_ring()
+        if R.cardinality() == Infinity:
+            raise NotImplementedError("iteration over infinite base ring not yet implemented")
+        iters = []
+        zero = R.zero()
+        v = []
+        n = 0
+        yield self.zero()
+        if R.is_zero():
+            return
+        while True:
+            if n == len(iters):
+                iters.append(iter(R))
+                v.append(next(iters[n]))
+                assert v[n] == zero, ("first element of iteration must be zero otherwise result "
+                                      "of this and free module __iter__ will be incorrect")
+            try:
+                v[n] = next(iters[n])
+                yield self(v)
+                n = 0
+            except StopIteration:
+                iters[n] = iter(R)
+                v[n] = next(iters[n])
+                assert v[n] == zero
+                n += 1
+
+    def cardinality(self):
+        """
+        Return the cardinality of the polynomial ring.
+
+        OUTPUT: either an integer or ``+Infinity``
+
+        EXAMPLES::
+
+            sage: R.<x> = ZZ[]
+            sage: R.cardinality()
+            +Infinity
+            sage: R.<x> = Integers(1)[]
+            sage: R.cardinality()
+            1
+        """
+        if self.base_ring().is_zero():
+            return 1
```

Comment:
```suggestion
            return Integer(1)
```
We try to avoid letting Python `int`s out into the wild.

Reply:
`ZZ.one()` is faster, guess we can use that instead.

```
sage: %timeit Integer(1)
149 ns ± 12.3 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
sage: %timeit ZZ.one()
83.5 ns ± 3.26 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Reply:
Either way is fine with me. I doubt this will be that much of a bottleneck, :P